src/operations/blas2/gemv.hpp

```diff
@@ -80,6 +80,244 @@ SYCL_BLAS_INLINE void AddSetColumns<rhs_t>::adjust_access_displacement() {
+   return true;
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t, bool is_transposed, int cache_line_size, int work_per_thread>
+SYCL_BLAS_INLINE
+typename Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::value_t
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::eval(cl::sycl::nd_item<1> ndItem) {
+
+  const index_t local_id = ndItem.get_local_id(0);
+  const index_t local_range = ndItem.get_local_range(0);
+  const index_t group_id = ndItem.get_group(0);
+  const index_t group_range = ndItem.get_group_range(0);
+  const index_t contract_dim = is_transposed ? matrix_.get_size_row() : matrix_.get_size_col();
+  const index_t non_contract_dim = is_transposed ? matrix_.get_size_col() : matrix_.get_size_row();
+  const index_t group_stride = group_range * local_range;
+  const index_t lda = matrix_.getSizeL();
+
+  const index_t work_load_per_thread = contract_dim;
+  const index_t thread_id = group_id * local_range + local_id;
+  if (thread_id >= non_contract_dim) {
+    return 0;
+  }
+
+  const index_t non_contract_local_thread_stride = is_transposed ? lda : 1;
+  const index_t contract_stride = is_transposed ?  1 : lda;
+  value_t sum = 0;
+  
+  for (index_t row_id = 0; row_id < non_contract_dim; row_id += group_stride) {
+
+    index_t non_contract_dim_index = (thread_id + row_id) * non_contract_local_thread_stride;
+
+    sum = 0;
+    for(index_t col_id = 0; col_id < contract_dim; ++col_id) { 
+      sum += ProductOperator::eval(matrix_. template eval<true>(non_contract_dim_index), vector_.eval(col_id));
+      non_contract_dim_index += contract_stride;
+    }
+    
+    lhs_.eval(thread_id + row_id) = sum;
+  }
+  
+  return 0;
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t, bool is_transposed, int cache_line_size, int work_per_thread>
+template <typename local_memory_t>
+SYCL_BLAS_INLINE
+typename Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::value_t
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::eval(local_memory_t local_mem, cl::sycl::nd_item<1> ndItem) {
+
+  const index_t local_id = ndItem.get_local_id(0);
+  const index_t local_range = ndItem.get_local_range(0);
+  const index_t group_id = ndItem.get_group(0);
+  const index_t group_range = ndItem.get_group_range(0);
+  const index_t lda = matrix_.getSizeL();
+  const index_t nc_dim = is_transposed ? matrix_.get_size_col() : matrix_.get_size_row();
+  const index_t c_dim = is_transposed ? matrix_.get_size_row() : matrix_.get_size_col();
+
+  // ID of the group in the non-contracting dimension of the global grid of WGs
+  const index_t nc_group_id = group_id / WGs_per_C_;
+  // ID of the group in the contracting dimension in the global grid of WGs
+  const index_t c_group_id = group_id % WGs_per_C_;
```

Comment:
Generally try to avoid the modulo in kernel code (except for compile-time expressions). This can be replaced with:
```
c_group_id = group_id - nc_group_id * WGs_per_C_
```

Reply:
Changed